src/Core/Dev/ServicesGraph/ServicesGraphDumper.php

```diff
@@ -0,0 +1,224 @@
+
+        // we tag all services to be extracted
+        foreach ($rootNodes as $serviceToExplore) {
+            $rootDefinition = $containerBuilder->getDefinition($serviceToExplore);
+            $this->tagDefinitionRecursivelyForGraphExploration($rootDefinition);
+        }
+
+        // we extract the tagged services
+        foreach ($containerBuilder->getDefinitions() as $id => $def) {
+            if ($def->hasTag('to-be-graphed')) {
+                $filteredContainer->setDefinition($id, $def);
+            }
+        }
+
+        return $filteredContainer;
+    }
+
+    /**
+     * @param ContainerBuilder $serviceContainer
+     *
+     * @return string
+     */
+    protected function dumpServiceContainer(ContainerBuilder $serviceContainer)
+    {
+        $dumper = new CustomGraphvizDumper($serviceContainer);
+        $graph = $dumper->dump();
+
+        return $graph;
+    }
+
+    /**
+     * Load the Symfony container with PrestaShop services
+     *
+     * @return ContainerBuilder
+     *
+     * @todo: be able to target PrestaShop services defined outside of PrestaShopBundle
+     */
+    protected function getPrestaShopContainerBuilder()
+    {
+        $container = new ContainerBuilder();
+
+        $loader = new YamlFileLoader($container, new FileLocator($this->prestashopBundleRootDirectory . '/Resources/config'));
+        $loader->load('services.yml');
+
+        return $container;
+    }
+
+    /**
+     * @param Definition $definition
+     * @param int $i
+     */
+    protected function tagDefinitionRecursivelyForGraphExploration(Definition $definition, $i = 0)
+    {
+        $i++;
+        $definition->addTag('to-be-graphed');
+
+        foreach ($definition->getArguments() as $argument) {
+            $this->tagDefinitionArgumentRecursively($argument, $i);
+        }
+    }
+
+    /**
+     * @param mixed $argument service definition argument
+     * @param int $i
+     */
+    protected function tagDefinitionArgumentRecursively($argument, $i)
+    {
+        if ($i > self::MAX_NESTED_SERVICES_SEARCH_DEPTH) {
+            throw new \RuntimeException('MAX_NESTED_SERVICES_DEFINITION_DEPTH reached, aborting...');
+        }
+
+        if ((is_string($argument)) || (is_array($argument))) {
+            return;
+        }
+        if (($argument instanceof Parameter) || ($argument instanceof \Symfony\Component\ExpressionLanguage\Expression)) {
+            // @todo: to be improved: this could be part of the graph
+            return;
+        }
+
+        if ($argument instanceof Reference) {
+            $serviceId = (string)$argument;
+            try {
+                $definition = $this->containerBuilder->getDefinition($serviceId);
+            } catch (ServiceNotFoundException $e) {
+                // @todo: to be improved: this could be part of the graph
+                return;
+            }
+            $this->tagDefinitionRecursivelyForGraphExploration($definition, $i);
```

Comment:
the tag is added twice: in `tagDefinitionRecursivelyForGraphExploration` and at the end of `tagDefinitionArgumentRecursively` (should be handled inside the `else`)
also is it logic to tag a service if it hasn't been found?